File: content/problems/2538-difference-between-maximum-and-minimum-price-sum/solutions/iterative-tree-dp-with-excluded-endpoint/Solution.java
class Solution {
    public long maxOutput(int n,int[][] edges,int[] price){
        List<Integer>[] graph=new ArrayList[n];
        for(int i=0;i<n;i++)graph[i]=new ArrayList<>();
        for(int[] edge:edges){
            graph[edge[0]].add(edge[1]);
            graph[edge[1]].add(edge[0]);
        }
        int[] parent=new int[n],order=new int[n];
        Arrays.fill(parent,-2);
        parent[0]=-1;
        int size=1;
        for(int index=0;index<size;index++){
            int node=order[index];
            for(int next:graph[node])if(parent[next]==-2){
                parent[next]=node;
                order[size++]=next;
            }
        }
        long[] full=new long[n],excluded=new long[n];
        long answer=0;
        for(int index=n-1;index>=0;index--){
            int node=order[index];
            long bestFull=price[node],bestExcluded=0;
            for(int child:graph[node])if(parent[child]==node){
                answer=Math.max(answer,Math.max(bestFull+excluded[child],bestExcluded+full[child]));
                bestFull=Math.max(bestFull,price[node]+full[child]);
                bestExcluded=Math.max(bestExcluded,price[node]+excluded[child]);
            }
            full[node]=bestFull;
            excluded[node]=bestExcluded;
        }
        return answer;
    }
}

File: content/problems/2538-difference-between-maximum-and-minimum-price-sum/solutions/iterative-tree-dp-with-excluded-endpoint/Solution.kt
class Solution {
    fun maxOutput(n:Int,edges:Array<IntArray>,price:IntArray):Long{
        val graph=Array(n){
            mutableListOf<Int>()
        };
        for(edge in edges){
            graph[edge[0]].add(edge[1]);
            graph[edge[1]].add(edge[0])
        };
        val parent=IntArray(n){
            -2
        };
        val order=IntArray(n);
        parent[0]=-1;
        var size=1;
        var index=0;
        while(index<size){
            val node=order[index++];
            for(next in graph[node])if(parent[next]==-2){
                parent[next]=node;
                order[size++]=next
            }
        };
        val full=LongArray(n);
        val excluded=LongArray(n);
        var answer=0L;
        for(position in n-1 downTo 0){
            val node=order[position];
            var bestFull=price[node].toLong();
            var bestExcluded=0L;
            for(child in graph[node])if(parent[child]==node){
                answer=maxOf(answer,bestFull+excluded[child],bestExcluded+full[child]);
                bestFull=maxOf(bestFull,price[node]+full[child]);
                bestExcluded=maxOf(bestExcluded,price[node]+excluded[child])
            };
            full[node]=bestFull;
            excluded[node]=bestExcluded
        };
        return answer
    }
}
